src/kernels/MIOpenCosineEmbeddingLoss.cpp

```diff
@@ -0,0 +1,400 @@
+/*******************************************************************************
+ *
+ * MIT License
+ *
+ * Copyright (c) 2024 Advanced Micro Devices, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#ifndef MIOPEN_DONT_USE_HIP_RUNTIME_HEADERS
+#include <hip/hip_fp16.h>
+#include <hip/hip_runtime.h>
+#endif
+
+#include "float_types.h"
+#include "tensor_view.hpp"
+
+#ifndef INPUT_TYPE
+#define INPUT_TYPE float
+#endif
+
+#ifndef OUTPUT_TYPE
+#define OUTPUT_TYPE float
+#endif
+
+#ifndef D_TYPE
+#define D_TYPE float
+#endif
+
+template <typename TI, typename TO>
+__device__ void cosineembeddinglossNorm2d(const TI* __restrict__ input1,
```

Comment:
Is there a reason why the originally one kernel was divided into two?

Reply:
The implementation method appears to have changed slightly. There seem to be pros and cons to the way you worked and the original method. Please compare moDNN with the current implementation. I think this work will take some time, so please postpone the due on the ticket.

Reply:
Actually, this kernel was divided into three kernels. In the first trial, I naively ported the Modnn kernel to a single MIOpen kernel, but the performance was poor in most cases, especially when D was large. The reason is that in Modnn, the kernel with N threads has a for loop that iterates over D:
```
for (size_t d = 0; d < D; d++) {                                          \
      cos_term += TV_2D_AT(I1, n, d) * TV_2D_AT(I2, n, d);                    \
      norm1 += TV_2D_AT(I1, n, d) * TV_2D_AT(I1, n, d);                       \
      norm2 += TV_2D_AT(I2, n, d) * TV_2D_AT(I2, n, d);                       \
    }    
```
To address this, I created a kernel named cosineembeddinglossNorm2d (1), which has N*D threads to calculate each cos_term, norm1, and norm2. Then, I applied the SumFwdContiguous kernel (2) to reduce these values along dimension 0, obtaining the final cos_term, norm1, and norm2. Finally, I made a kernel to perform the final calculations as in the Modnn kernel (3).

I compared the performance of the naive version and the optimized version in MIOpen, and the optimized version performed better in average.

Do I still need to compare the current implementation with the Modnn version?

Reply:
I already understand your implementation, but I don't think the kernel time measurement is right now. Please look at https://github.com/moreh-dev/MIOpen/pull/22#discussion_r1621748562 and measure the op time.